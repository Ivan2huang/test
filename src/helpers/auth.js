export const getCookie = name => {
  const value = `; ${document.cookie}`;
  const parts = value.split(`; ${name}=`);
  return parts
    .pop()
    .split(';')
    .shift();
};

export const hasCookie = name => {
  const value = getCookie(name);
  return value && value.trim().length > 0;
};
