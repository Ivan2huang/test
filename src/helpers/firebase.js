import firebase from 'firebase/app';
import 'firebase/analytics';
import 'firebase/remote-config';
import CONFIG from '../constants/config';
import logger from './logger';

const config = {
  apiKey: CONFIG.firebaseApiKey,
  authDomain: CONFIG.firebaseAuthDomain,
  databaseURL: CONFIG.firebaseDatabaseUrl,
  projectId: CONFIG.firebaseProjectId,
  storageBucket: CONFIG.firebaseStorageBucket,
  messagingSenderId: CONFIG.firebaseMessagingSenderId,
  appId: CONFIG.firebaseAppId,
  measurementId: CONFIG.firebaseMeasurementId,
};

let firebaseApp;

export const initFirebase = () => {
  if (!firebaseApp) {
    firebaseApp = firebase.initializeApp(config);
  } else {
    firebaseApp = firebase.app();
  }
};

export const trackUserInfo = ({ clientId, memberId, role }) => {
  const analytics = firebaseApp.analytics();
  analytics.setUserId(memberId);
  analytics.setUserProperties({
    client_id: clientId,
    user_type: role,
  });
};

export const logEvent = (eventName, category, action, label, options = {}) => {
  if (CONFIG.enableGA) {
    firebaseApp.analytics().logEvent(eventName, {
      category,
      action,
      label,
      ...options,
    });
  }
};

export const logAction = ({ category, action, ...restParams }) => {
  if (CONFIG.enableGA) {
    firebaseApp.analytics().logEvent('user_action', {
      category,
      action,
      ...restParams,
    });
  }
};

export const trackingAnchor = e => {
  e.preventDefault();
  const anchorNode = e.target;
  const href = anchorNode.getAttribute('href');
  const { category, action } = anchorNode.dataset;
  logAction({
    category,
    action,
  });
  window.location.assign(href);
};

export const getFirebaseApp = () => firebaseApp;

export const getRemoteFeatureChecker = async ({ parameterName }) => {
  if (!firebaseApp) {
    return () => false;
  }

  const remoteConfig = firebaseApp.remoteConfig();
  remoteConfig.settings = {
    minimumFetchIntervalMillis: 3600000,
  };
  let featureMap = {};
  try {
    await remoteConfig.fetchAndActivate();
    const remoteValues = await remoteConfig.getAll();
    if (remoteValues && remoteValues[parameterName]) {
      featureMap = JSON.parse(remoteValues[parameterName].asString());
    }
  } catch (e) {
    logger.error('---getRemoteFeatureChecker -> ERROR---', e);
  }
  return featureName => {
    return featureMap[featureName];
  };
};
