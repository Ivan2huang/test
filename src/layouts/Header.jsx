import React from 'react';
import { connect } from 'react-redux';
import { compose } from 'redux';
import PropTypes from 'prop-types';
import { Box, withStyles } from '@material-ui/core';
import NavBar from './NavBar';
import PageLoader from './pageLoader';
import { loaderDetail } from '../modules/loader/util';
import LOADER from '../constants/loader';
import HeaderBrand from '../uiComponents/HeaderBrand';

export const style = theme => ({
  boxRoot: {
    position: 'fixed',
    zIndex: '10000',
    width: '100%',
    minHeight: theme.spacingX(20),
    display: 'flex',
    justifyContent: 'space-between',
    alignItems: 'center',
    background: theme.headerBackground || theme.highEmphasis,
    boxShadow: `0 1px 0 ${theme.grey1}`,
    marginBottom: theme.spacingX(40),
    [theme.breakpoints.down('sm')]: {
      paddingRight: theme.spacingX(10),
      paddingLeft: theme.spacingX(8),
      minHeight: theme.spacingX(14),
    },
  },
  loaderContainer: {
    zIndex: '10001',
    position: 'fixed',
    left: '48%',
    top: '20px',
    [theme.breakpoints.down('sm')]: {
      top: '10px',
      left: '45%',
    },
  },
});

const mapStateToProps = ({ loader }) => ({
  ...loaderDetail(loader, LOADER.page),
});

const Header = ({ classes, loading }) => {
  return (
    <>
      <Box component="header">
        <Box classes={{ root: classes.loaderContainer }}>
          <PageLoader loading={loading} />
        </Box>
        <Box classes={{ root: classes.boxRoot }}>
          <HeaderBrand />
          <NavBar />
        </Box>
      </Box>
    </>
  );
};
Header.defaultProps = {
  loading: false,
};

Header.propTypes = {
  classes: PropTypes.exact({
    boxRoot: PropTypes.string.isRequired,
    loaderContainer: PropTypes.string.isRequired,
  }).isRequired,
  loading: PropTypes.bool,
};

export default compose(
  connect(mapStateToProps),
  withStyles(style),
)(Header);
