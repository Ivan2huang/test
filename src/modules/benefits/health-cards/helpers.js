/* eslint-disable import/prefer-default-export */
export const getQRCodeXML = ({
  fullName,
  membershipNumber,
  certificateNumber,
  insurerCode,
  insurerName,
  insurerType,
  policyNumber,
  expiryDate,
}) =>
  [
    '<?xml version="1.0"?>',
    '<medical_card>',
    '<Card_type>',
    insurerType,
    '</Card_type>',
    '<insurance_company_code>',
    insurerCode,
    '</insurance_company_code>',
    '<insurance_company_name>',
    insurerName,
    '</insurance_company_name>',
    '<insured_name>',
    fullName,
    '</insured_name>',
    '<policy_no>',
    policyNumber,
    '</policy_no>',
    '<certificate_no>',
    certificateNumber,
    '</certificate_no>',
    '<membership_no>',
    membershipNumber,
    '</membership_no>',
    '<expiry_date>',
    expiryDate,
    '</expiry_date>',
    '</medical_card>',
  ].join('');
