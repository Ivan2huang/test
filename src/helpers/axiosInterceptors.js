/* eslint-disable import/no-cycle */
import url from './url';
import paths from './paths';
import { error as errorCodes } from '../constants/auth';

let requestsToRefresh = [];
let isRefreshRequesting = false;

export default axios => {
  if (axios) {
    axios.interceptors.response.use(
      response => {
        return response;
      },
      err => {
        const { response, config } = err;

        if ([401, 403].indexOf(response.status) !== -1) {
          if (!isRefreshRequesting) {
            isRefreshRequesting = true;
            axios
              .get(url.refreshToken)
              .then(({ data }) => {
                const { isSuccess, error } = data;
                if (isSuccess) {
                  requestsToRefresh.forEach(cb => cb(true));
                } else {
                  const redirectUrl =
                    error === errorCodes.token.invalidRefreshToken
                      ? paths.common.tokenExpired
                      : paths.common.login;
                  window.location.replace(redirectUrl);
                }
              })
              .catch(() => {
                requestsToRefresh.forEach(cb => cb(false));
              })
              .finally(() => {
                requestsToRefresh = [];
                isRefreshRequesting = false;
              });
          }

          return new Promise((resolve, reject) => {
            requestsToRefresh.push(hasToken => {
              if (hasToken) {
                resolve(axios(config));
              } else {
                reject(Promise.reject(err));
              }
            });
          });
        }

        return Promise.reject(err);
      },
    );
  }
};
