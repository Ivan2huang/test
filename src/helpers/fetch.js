/* eslint-disable import/no-cycle */
/* eslint-disable no-console,consistent-return */

import axios from 'axios';
import { navigateTo } from './helpers';
import paths from './paths';
import useAxiosInterceptors from './axiosInterceptors';
import getLocale from '../i18n/getLocale';
import CONFIG from '../constants/config';
import logger from './logger';

if (CONFIG.enableRefreshToken) {
  useAxiosInterceptors(axios);
}

export const ContentTypes = {
  JSON: 'application/json;charset=UTF-8',
  FILE: 'application/x-www-form-urlencoded',
};

export const defaultHeader = () => ({
  accept: ContentTypes.JSON,
  'Content-Type': ContentTypes.JSON,
  'Accept-Language': getLocale(),
  Pragma: 'no-cache',
});

const uploadFileHeader = () => ({
  'Content-Type': ContentTypes.FILE,
});

const handleError = (error, expectsErrorResponse) => {
  let response = false;
  logger.error('fetch -> ERROR', error);
  if (error.response && error.response.data) {
    response =
      error.response.data.errors && error.response.data.errors.length >= 1
        ? {
            error: true,
            ...error.response.data.errors[0],
          }
        : {
            error: true,
            messageKey: error.response.data.code,
            message: error.response.data.message,
          };
  } else if (error && !error.status) {
    // network error
    response = {
      error: true,
      messageKey: 'NETWORK_ERROR',
      message: '',
    };
  }

  if (expectsErrorResponse) {
    return response;
  }

  if (error.response && error.response.status === 404) {
    navigateTo(paths.common.notfound);
  } else {
    navigateTo(paths.common.error);
  }
};

export const fetchData = (
  method,
  url,
  body,
  expectsErrorResponse = false,
  headerOptions = {},
) => {
  return axios(url, {
    headers: {
      ...defaultHeader(),
      ...headerOptions,
    },
    method,
    data: JSON.stringify(body),
  })
    .then(response => response.data)
    .catch(error => handleError(error, expectsErrorResponse));
};

export const uploadFile = (method, url, body, expectsErrorResponse = false) => {
  return axios(url, {
    headers: uploadFileHeader(),
    method,
    data: body,
  })
    .then(response => response.data)
    .catch(error => handleError(error, expectsErrorResponse));
};

export const fetchFile = url => {
  const axiosOption = { responseType: 'blob' };
  return axios(url, axiosOption)
    .then(response => response.data)
    .catch(error => handleError(error, false));
};
