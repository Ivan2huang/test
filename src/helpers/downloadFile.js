import { fetchFile } from './fetch';

const SUCCESS = 'SUCCESS';
const ERROR = 'ERROR';

const downloadFileAsync = async (url, fileName) => {
  const response = await fetchFile(url);
  if (response) {
    const blob = new Blob([response]);
    if (window.navigator.msSaveOrOpenBlob) {
      window.navigator.msSaveOrOpenBlob(blob, fileName);
      return SUCCESS;
    }
    const objectUrl = window.URL.createObjectURL(blob);
    const link = document.createElement('a');
    link.href = objectUrl;
    link.setAttribute('download', fileName);
    document.body.appendChild(link);
    link.click();
    document.body.removeChild(link);
    return SUCCESS;
  }
  return ERROR;
};

export default downloadFileAsync;
