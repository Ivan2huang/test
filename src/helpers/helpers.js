import Router from 'next/router';

import zhLocale from 'date-fns/locale/zh-CN';
import enLocale from 'date-fns/locale/en-US';
import * as Sentry from '@sentry/node';

import DateFnsUtils from '@date-io/date-fns';
import { getCookie } from './auth';
import { USER_ID } from '../constants/auth';
import CONFIG from '../constants/config';
// eslint-disable-next-line import/no-cycle
import { fetchFile } from './fetch';

const PRECISION = 2;

export const LOCALES_MAP = {
  'en-HK': enLocale,
  'zh-HK': zhLocale,
};

export const formatMessage = (intl, id, defaultMessage, values) =>
  intl.formatMessage({ id, defaultMessage }, values);

export const maskEmail = email => {
  const [localPart, domain] = email.split('@');
  return `${localPart[0]}******${localPart[localPart.length - 1]}@${domain}`;
};

export const formatAmount = (intl, amount) =>
  intl.formatNumber(amount, {
    minimumFractionDigits: PRECISION,
    maximumFractionDigits: PRECISION,
  });

export const navigateTo = (path, queryParam = {}, asPath) => {
  console.log('helpers.js navigateTo path=', path);
  Router.push({ pathname: path, query: queryParam }, asPath);
};

export const buildClaimPostBody = (
  formData,
  receiptFilesIds,
  referralFileId,
) => {
  const selectedPatientId = formData.patient.patientId;
  const userId = getCookie(USER_ID);
  const claimantName =
    selectedPatientId === userId
      ? `Employee|${selectedPatientId}`
      : `Dependant|${selectedPatientId}`;

  return {
    benefitPeriod: formData.planId,
    claimantName,
    receiptDate: formData.claim.consultationDate,
    receiptAmount: formData.claim.receiptAmount,
    claimReason: formData.claim.diagnosis,
    claimTypeId: formData.claim.claimId,
    claimItemName: formData.claimType,
    int1: true,
    str5: formData.patient.contactNumber,
    receiptFilesIds,
    referralFileId,
    int2: formData.claim.anotherInsurer ? 1 : 0,
    dec1: formData.claim.otherInsurerAmount,
  };
};

const isValidFileType = (validFileType, fileName) =>
  validFileType.includes(
    fileName
      .split('.')
      .pop()
      .toLowerCase(),
  );

const isValidFileSize = (validFileSize, fileSize) => fileSize <= validFileSize;

export const validateFiles = (files, validFileType, validFileSize) => {
  return files.filter(
    ({ name, size }) =>
      isValidFileType(validFileType, name) &&
      isValidFileSize(validFileSize, size),
  );
};

export const isEmpty = obj =>
  Object.entries(obj).length === 0 && obj.constructor === Object;

export const today = () =>
  new DateFnsUtils().format(Date.now(), CONFIG.dateFormat);

export const formatDate = (
  date,
  locale = null,
  customFormat = CONFIG.dateFormat,
) => {
  try {
    const dateFnsUtil = new DateFnsUtils();
    if (date && date.includes('0001')) {
      return undefined;
    }
    if (locale) {
      dateFnsUtil.locale =
        LOCALES_MAP[locale] || LOCALES_MAP[CONFIG.locales.ENGLISH];
    }
    return dateFnsUtil.format(new Date(date), customFormat);
  } catch (e) {
    return undefined;
  }
};

export const getCurrentDate = () => new Date();

export const addDays = days => new DateFnsUtils().addDays(new Date(), days);

export const openFile = async (name, url, preloaded = true) => {
  if (window.navigator && window.navigator.msSaveOrOpenBlob) {
    if (!preloaded) {
      window.open(url);
      return;
    }

    const blob = await fetchFile(url);
    window.navigator.msSaveOrOpenBlob(blob, name);
    return;
  }

  window.open(url);
};
export const onEnter = (event, callback) => {
  if (callback && event.key === 'Enter') {
    callback();
  }
};

export const saveToLocalStorage = (key, value) => {
  window.localStorage.setItem(key, value);
};

export const getValueFromLocalStorage = key => {
  return window.localStorage.getItem(key);
};

export const sentenceCase = input => {
  return input.replace(/\w\S*/g, word => {
    return word.charAt(0).toUpperCase() + word.substr(1).toLowerCase();
  });
};

export const delay = waitFor =>
  new Promise(resolve => {
    setTimeout(() => resolve(), waitFor);
  });

export const objectToURLParams = obj =>
  Object.keys(obj)
    .map(key => {
      return `${key}=${encodeURIComponent(obj[key])}`;
    })
    .join('&');

export const snakeToCamel = str =>
  str.replace(/([-_]\w)/g, g => g[1].toUpperCase());

export const objectKeySnakeToCamel = obj => {
  return Object.entries(obj).reduce((result, [key, value]) => {
    return {
      ...result,
      [snakeToCamel(key)]: value,
    };
  }, {});
};

export const isValidLanguageCode = code => {
  return (CONFIG.supportedLanguages || []).includes(code);
};

/* istanbul ignore next */
export const captureException = (error, errorInfo = {}) => {
  Sentry.withScope(scope => {
    Object.keys(errorInfo).forEach(key => {
      scope.setExtra(key, errorInfo[key]);
    });

    if (error.response && error.response.data) {
      scope.setExtra(
        'requestErrors',
        JSON.stringify(error.response.data.errors || []),
      );
    }

    Sentry.captureException(error);
  });
};

export const removeUnicodeScript = str => {
  if (!str) return '';
  const pattern = /[\u006E\u00B0\u00B2\u00B3\u00B9\u02AF\u0670\u0711\u2121\u213B\u2207\u29B5\uFC5B-\uFC5D\uFC63\uFC90\uFCD9\u2070\u2071\u2074-\u208E\u2090-\u209C\u0345\u0656\u17D2\u1D62-\u1D6A\u2A27\u2C7C]+\s/;
  return str.replace(pattern, '').trim();
};

export const convertObjectToParam = (
  obj,
  compareSign = '=',
  delimiter = '&',
) => {
  if (typeof obj === 'object') {
    return Object.keys(obj)
      .map(key => `${key}${compareSign}${obj[key]}`)
      .join(delimiter);
  }
  return '';
};

export const lowercaseObjectKeys = obj => {
  if (typeof obj !== 'object') return {};

  const result = {};
  Object.keys(obj).forEach(key => {
    result[key.toLowerCase()] = obj[key];
  });

  return result;
};
