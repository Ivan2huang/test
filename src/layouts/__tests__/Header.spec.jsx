import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import { withStyles } from '@material-ui/core';

import Header, { style } from '../Header';
import withTheme from '../../themes/withThemeProvider';
import { navigateTo } from '../../helpers/helpers';
import withRedux from '../../redux/withReduxProvider';
import withIntl from '../../i18n/withIntlProvider';

jest.mock('../NavBar', () => props => <div {...props}>NavBar Component</div>);
jest.mock('../../helpers/helpers', () => ({
  navigateTo: jest.fn(),
  isEmpty: jest.fn(),
}));
jest.mock('../../constants/config', () => ({
  locales: {
    ENGLISH: 'en-HK',
    CHINESE: 'zh-HK',
  },
  defaultLanguage: 'en-HK',
}));

describe('Header Component', () => {
  it('should match snapshot', () => {
    const Component = withRedux(withTheme(withStyles(style)(withIntl(Header))));
    const { container } = render(<Component />);

    expect(container).toMatchSnapshot();
  });

  it('should navigate to default on click of Logo', () => {
    const Component = withRedux(withTheme(withStyles(style)(withIntl(Header))));
    const { getByTestId } = render(<Component />);
    const linkLogo = getByTestId('link-logo');
    fireEvent.click(linkLogo);
    expect(navigateTo).toHaveBeenCalledWith('/');
  });

  it('should use default logo when selected logo with language not found', () => {
    const Component = withRedux(withTheme(withStyles(style)(withIntl(Header))));
    const { container } = render(<Component preferredLocale="en-US" />);

    expect(container).toMatchSnapshot();
  });
});
