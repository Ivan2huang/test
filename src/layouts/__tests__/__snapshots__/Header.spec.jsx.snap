// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header Component should match snapshot 1`] = `
<div>
  <header
    class="MuiBox-root MuiBox-root-6"
  >
    <div
      class="MuiBox-root MuiBox-root-7 Header-loaderContainer-4 _class-loaderContainer-2"
      classes="[object Object]"
    >
      <div
        class="ScreenReader-srOnly-8"
      >
        <div
          aria-atomic="true"
          aria-live="assertive"
          class="MuiTypography-root makeStyles-root-9 makeStyles-root-10 MuiTypography-body1"
          role="alert"
        />
      </div>
    </div>
    <div
      class="MuiBox-root MuiBox-root-41 Header-boxRoot-3 _class-boxRoot-1"
      classes="[object Object]"
    >
      <button
        class="MuiBox-root MuiBox-root-46 HeaderBrand-imgContainer-42 HeaderBrand-imgContainer-44"
        classes="[object Object]"
        data-testid="link-logo"
        tabindex="0"
      >
        <img
          alt=""
          class="HeaderBrand-img-43 HeaderBrand-img-45"
          src="/static/logo.en.svg"
        />
      </button>
      <div>
        NavBar Component
      </div>
    </div>
  </header>
</div>
`;

exports[`Header Component should use default logo when selected logo with language not found 1`] = `
<div>
  <header
    class="MuiBox-root MuiBox-root-98"
  >
    <div
      class="MuiBox-root MuiBox-root-99 Header-loaderContainer-96 _class-loaderContainer-94"
      classes="[object Object]"
    >
      <div
        class="ScreenReader-srOnly-100"
      >
        <div
          aria-atomic="true"
          aria-live="assertive"
          class="MuiTypography-root makeStyles-root-101 makeStyles-root-102 MuiTypography-body1"
          role="alert"
        />
      </div>
    </div>
    <div
      class="MuiBox-root MuiBox-root-133 Header-boxRoot-95 _class-boxRoot-93"
      classes="[object Object]"
    >
      <button
        class="MuiBox-root MuiBox-root-138 HeaderBrand-imgContainer-134 HeaderBrand-imgContainer-136"
        classes="[object Object]"
        data-testid="link-logo"
        tabindex="0"
      >
        <img
          alt=""
          class="HeaderBrand-img-135 HeaderBrand-img-137"
          src="[object Object]"
        />
      </button>
      <div>
        NavBar Component
      </div>
    </div>
  </header>
</div>
`;
